api/v3/Case.php

```diff
@@ -60,58 +62,112 @@
  *   api result array
  */
 function civicrm_api3_case_create($params) {
+  _civicrm_api3_case_format_params($params);
 
-  if (!empty($params['id'])) {
-    return civicrm_api3_case_update($params);
+  if (empty($params['id'])) {
+    // Creating a new case, so make sure we have the necessary parameters
+    civicrm_api3_verify_mandatory($params, NULL, array(
+        'contact_id',
+        'subject',
+        array('case_type', 'case_type_id'),
+      )
+    );
   }
+  else {
+    // Update an existing case
+    // FIXME: Some of this logic should move to the BAO object?
+    // FIXME: Should we check if case with ID actually exists?
+    if (!isset($params['case_id']) && isset($params['id'])) {
+      $params['case_id'] = $params['id'];
+    }
 
-  civicrm_api3_verify_mandatory($params, NULL, array(
-    'contact_id',
-    'subject',
-    array('case_type', 'case_type_id'))
-  );
-  _civicrm_api3_case_format_params($params);
+    if (array_key_exists('creator_id', $params)) {
+      throw new API_Exception('You cannot update creator id');
+    }
 
-  // If format_params didn't find what it was looking for, return error
-  if (empty($params['case_type_id'])) {
-    throw new API_Exception('Invalid case_type. No such case type exists.');
-  }
-  if (empty($params['case_type'])) {
-    throw new API_Exception('Invalid case_type_id. No such case type exists.');
-  }
+    $mergedCaseId = $origContactIds = array();
 
-  // Fixme: can we safely pass raw params to the BAO?
-  $newParams = array(
-    'case_type_id' => $params['case_type_id'],
-    'creator_id' => $params['creator_id'],
-    'status_id' => $params['status_id'],
-    'start_date' => $params['start_date'],
-    'end_date' => CRM_Utils_Array::value('end_date', $params),
-    'subject' => $params['subject'],
-  );
+    // get original contact id and creator id of case
+    if (!empty($params['contact_id'])) {
+      // FIXME: CRM_Case_BAO_Case::retrieveContactIdsByCaseId returns a 1-based array (1 is the first element)
+      // It should really return a 0-based array for consistency.
+      $origContactIds = CRM_Case_BAO_Case::retrieveContactIdsByCaseId($params['id']);
+      $origContactId = $origContactIds[1];
```

Comment:
You could use `CRM_Utils_Array::first()` so you don't have to worry about how the array is indexed.